signac/contrib/job.py

```diff
@@ -24,31 +23,170 @@
 logger = logging.getLogger(__name__)
 
 
-class _sp_save_hook:
-    """Hook to handle job migration when state points are changed.
-
-    When a job's state point is changed, in addition
-    to the contents of the file being modified this hook
-    calls :meth:`~Job._reset_sp` to rehash the state
-    point, compute a new job id, and move the folder.
-
-    Parameters
-    ----------
-    jobs : iterable of `Jobs`
-        List of jobs(instance of `Job`).
+class _StatepointDict(JSONDict):
+    """A JSON-backed dictionary for storing job statepoints.
 
+    There are three principal reasons for extending the base JSONDict:
+        1. Saving needs to trigger a job directory migration, and
+        2. Statepoints are assumed to not support external modification, so
+           they never need to load from disk _except_ the very first time a job
+           is opened by id and they're not present in the cache.
+        3. It must be possible to load and/or save on demand during tasks like
+           job directory migrations.
     """
 
-    def __init__(self, *jobs):
-        self.jobs = list(jobs)
+    _PROTECTED_KEYS = ("_jobs",)
+
+    def __init__(
+        self,
+        jobs=None,
+        filename=None,
+        write_concern=False,
+        data=None,
+        parent=None,
+        *args,
+        **kwargs,
+    ):
+        # A job-statepoint mapping need not be unique because multiple Python
+        # Job objects can point to the same data on disk. We need to store
+        # these jobs in a shared list here so that shallow copies can point to
+        # the same place and trigger each other to update.
+        self._jobs = list(jobs)
```

Comment:
Nice idea, I hadn't considered that possibility.

Reply:
Although this is a good idea in principle, it's unfortunately not so easy in practice because 1) weakrefs are not pickleable by default, and 2) the interaction between weakrefs and the copy module has some gotchas, a simple replacement as suggested here does not lead to the `_jobs` being correctly modified by setstate (but it works fine with a list). We could probably make it work by implementing multiple additional methods, but it would require substantial additional code for no practical benefit, just some nice semantic clarity, so I'm going to leave this as is.